crates/ruff_linter/src/rules/flake8_pie/rules/reimplemented_container_builtin.rs

```diff
@@ -0,0 +1,91 @@
+use ruff_python_ast::{self as ast, Expr, ExprLambda};
+
+use ruff_diagnostics::{Diagnostic, Edit, Fix};
+use ruff_diagnostics::{FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for lambdas that can be replaced with the `list` builtin.
+/// If [preview] mode is enabled, then we will also look for lambdas
+/// that can be replaced with the `dict` builtin.
+///
+/// ## Why is this bad?
+/// Using container builtins is more readable.
+///
+/// ## Example
+/// ```python
+/// from dataclasses import dataclass, field
+///
+///
+/// @dataclass
+/// class Foo:
+///     bar: list[int] = field(default_factory=lambda: [])
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from dataclasses import dataclass, field
+///
+///
+/// @dataclass
+/// class Foo:
+///     bar: list[int] = field(default_factory=list)
+///     baz: dict[str, int] = field(default_factory=dict)
+/// ```
+///
+///
+/// If [preview]
+/// ## References
+/// - [Python documentation: `list`](https://docs.python.org/3/library/functions.html#func-list)
+///
+/// [preview]: https://docs.astral.sh/ruff/preview/
+#[violation]
+pub struct ReimplementedContainerBuiltin(&'static str);
```

Comment:
I could make this an `enum` instead, but not sure it's worthwhile compared to just storing the builtin inline.

Reply:
I decided to make it an enum to follow the pattern we use elsewhere, even though it's really unimportant.